File: Problem_2.java
// Time Complexity : O(n) where n is the length of the input
// Space Complexity : O(k) where k is the number of distinct messages
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no

// Your code here along with comments explaining your approach

import java.util.HashMap;

class Logger {
    HashMap<String, Integer> map;

    public Logger() {
        this.map = new HashMap<>();
    }

    public boolean shouldPrintMessage(int timestamp, String message) {
        // Check if the message is present in the map
        if (this.map.containsKey(message)) {
            // Check the timestamp associated with map
            if (timestamp >= this.map.get(message)) {
                this.map.put(message, timestamp + 10);
                return true;
            } else
                return false;
        } else {
            this.map.put(message, timestamp + 10);
            return true;
        }
    }
}

/**
 * Your Logger object will be instantiated and called as such:
 * Logger obj = new Logger();
 * boolean param_1 = obj.shouldPrintMessage(timestamp,message);
 */

File: problem2.cpp
// Time Complexity :O(nl) length of input string * average length of each dictionary
// Space Complexity :O(n) lenght of the input string
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this :


// Your code here along with comments explaining your approach
/*
*/



#include<iostream>
#include<vector>
#include<string>

using namespace std;

class Solution {
    vector<string> res;
    vector<string> hash_map{
        "abc",
        "def",
        "ghi",
        "jkl",
        "mno",
        "pqrs",
        "tuv",
        "wxyz"
    };
    void helper(string& digits,int level,string& s){
        //base case
        if(level == digits.length()){
            string s_push(s);
            res.push_back(s_push);
            return;
        }

        int hash_id = digits.at(level) - '2';
        string hash_val = hash_map[hash_id];

        for(int i{};i<hash_val.length();++i){
            s.push_back(hash_val.at(i));
            helper(digits,level+1,s);
            s.pop_back();
        }
    }
public:
    vector<string> letterCombinations(string digits) {
        if(digits == "") return res;
        string s{""};
        helper(digits,0,s);
        return res;
    }
};